### ISSUE TITLE
How to get a State from an Effect

### ISSUE THREAD

**Issue by subgurim** [mrpmorris/blazor-fluxor#113]
First of all thanks for this Incredible library!

Scenary:
- There are two States: StateA and StateB.
- ActionA is dispatched and catched by ReducerA and EffectA

Question:
- Can StateB be accessed from EffectA? The current StateB, of course there's no need to listen to changes of the state.
   - I've tried to inject IState<StateB> on EffectB but doesn't work properly
   - StateB cannot be injected (the states have their own constructors)

What I'm doing now is:
- Fill ActionA with values of StateB (which is not maintainable if I need N states)

Thanks in advance!!

**Comment by subgurim** [comment]
Hi @mrpmorris, any idea about this? A lot of thanks!!

**Comment by mrpmorris** [comment]
Sorry, I didn't see this question.

Injecting state into effects is an unusual practice. Can you tell me what it is you need to achieve? Perhaps there is another way.

**Comment by subgurim** [comment]
Of course @mrpmorris ! Thanks for your attention.

I'm working on a shop that package three different kind of products: tickets, hotels and car rental.

Each one has it's own sources, so I have One State for each one:
- TicketsState
- HotelsState
- CarRentalState

Every sale process has 3 basic steps:
1.- Get the product
2.- Reserve the product
3.- Purchase the product

The first step is simple: Tickets, Hotels and CarRental have their own source so I get the product using their own Effect (works perfect).

The second and third steps are not that simple. Focusing on second step (third is analogous) I have a "ReservationState" with its CreateReservationAction --> CreateReservationReducer --> CreateReservationEffect --> CreateReservationSucceedAction --> CreateReservationSucceedReducer --> CreateReservationSucceedEffect... and here is the "problem".

CreateReservationEffect (and CreateReservationSucceedEffect) needs TicketsState, HotelsState, CarRentalState. Getting products are independent, but reserving products are a "on call thing". (This scenario is extremely simplified).

So I'm setting the 3 states in CreateReservationAction and then to CreateReservationSucceedAction... but it would be extremely more simple if I could get TicketsState, HotelsState, CarRentalState in CreateReservationEffect/CreateReservationSucceedEffect.

In other words: I want to change ReservationState and, as a direct consecuence I want to use TicketState, HotelsState and CarRentalState.


I hope I explained myself well :)

Thanks!!!

**Comment by mrpmorris** [comment]
I think your action should contain all the data you need. The process (UI) that creates the action has access to any state it needs. Create the relevant action, set all the data you will require, then Dispatch that action. The effect will then have all the data it needs in the action it is passed.

If you need to, when your effect posts to the server it can use data from the result + data from the initial action that triggered the effect when creating a new action to Dispatch. For example, your initial action has IDs from 3 states, the effect goes to the server, the effect then creates a new action that has the IDs in the action + data from the server response, and then dispatches the new response.

Does that help?

PS: Instead of XXXXSucceededAction and XXXXFailedAction, you can just have a XXXCompletedAction with a "bool Succeeded" property - if you prefer.

**Comment by subgurim** [comment]
Hi @mrpmorris,

Thanks for your advices. At least I know I'm doing it well (the initial action has all the needed data).

A lot of thanks for your help!

Best Regards.
